Prow secrets management docu (#51)
kyma-project/test-infra: prow-configurations/prow-secrets-management.md

# Prow secrets management

## Overview

This document explains secrets management for Prow. Some jobs require using sensitive data. You need to encrypt data using KMS and store them in GCS. The following instructions assume that you are logged-in to the Google Cloud project with administrative rights.

## Prerequisities

>  - [gcloud](https://cloud.google.com/sdk/gcloud/) installed. 

Reviewer: ```suggestion
 - [gcloud](https://cloud.google.com/sdk/gcloud/) 
```

 - [gcloud](https://cloud.google.com/sdk/gcloud/) 
 - basic knowledge about [GCP key rings and keys](https://cloud.google.com/kms/docs/creating-keys).

For your convenience, export these data:
 - PROJECT_NAME - Google Cloud Project
 - BUCKET_NAME - GCS bucket in the $PROJECT_NAME where Prow secrets are stored
 - KEYRING_NAME - KMS key ring
 - ENCRYPTION_KEY_NAME - key in the $KEYRING_NAME, used for data encryption

> 

Reviewer 2: ```
Use the `export {VARIABLE}={value}` command to set up these variables, where:
 - **PROJECT_NAME** is a Google Cloud project.
 - **BUCKET_NAME** is a GCS bucket in the Google Cloud project that is used to store Prow secrets.
 - **KEYRING_NAME** is the KMS key ring.
 - **ENCRYPTION_KEY_NAME** is the key name in the key ring that is used for data encryption.
```

> ->**NOTE:** Before following this guide check Prow secrets setup for the $PROJECT_NAME.

Reviewer 3: ```suggestion
->**NOTE:** Perform each step listed below only if not done already.
```

> >**NOTE:** Before following this guide, check Prow secrets setup for the $PROJECT_NAME.

Reviewer 2: - Before following -> Before you follow
- for the $PROJECT_NAME. -> for the Google Cloud project.

> Set context to the $PROJECT_NAME executing:

Reviewer 2: Execute this command to set the context to the Google Cloud project:

gcloud config set project $PROJECT_NAME
```

### Create GCS bucket

> 

Reviewer 3: ```suggestion

The purpose of the bucket is to store encrypted credentials necessary for prow jobs like provisioning clusters/vms etc.
```

> Run following command:

Reviewer 3: ```suggestion
Run the following command:
```

> ### Create GCS bucket

Reviewer 2: Create GCS bucket -> Create a GCS bucket

> The purpose of the bucket is to store encrypted credentials necessary for prow jobs like provisioning clusters/vms etc.

Reviewer 2: - prow -> Prow
- like provisioning clusters/vms etc. -> like provisioning clusters or virtual machines.

> Run this command:

Reviewer 2: Run this command to create a bucket:

```
gsutil mb -p $PROJECT_NAME gs://$BUCKET_NAME/
```

> ### Create Google Service Account

Reviewer 2: Create Google Service Account - Create a Google service account

Export these variables:
 - SA_NAME - Service Account name
 - SA_DISPLAY_NAME - Service Account display name
 - SECRET_FILE - the path where the private key will be written
 - ROLE - the role bound to the $SA_NAME

> 

Reviewer 2: ```
Export the variables, where:
 - **SA_NAME** is the name of the service account.
 - **SA_DISPLAY_NAME** is the display name of the service account.
 - **SECRET_FILE** is the path for the private key.
 - **ROLE** is the role bound to the service account.

```

Create a service account:
```
gcloud iam service-accounts create $SA_NAME --display-name $SA_DISPLAY_NAME
```

> Create a private key for the $SA_NAME service account:

Reviewer 2: for the $SA_NAME service account: -> for the service account:

```
gcloud iam service-accounts keys create $SECRET_FILE --iam-account=SA_NAME
```

> Add policy binding for the $SA_NAME service account:

Reviewer 2: Add a policy binding for the service account:

```
gcloud iam service-accounts add-iam-policy-binding $SA_NAME --member=serviceAccount:$SA_NAME@$PROJECT_NAME.iam.gserviceaccount.com --role=$ROLE
```

> ### Encrypt the secret

Reviewer 2: secret -> Secret

> Export the following:

Reviewer 2: Export the **SECRET_FILE** variable which is the path to the file containing the secret.

 - SECRET_FILE - path to the file containing secret

Encrypt the secret:
```
gcloud kms encrypt --location global --keyring $KEYRING_NAME --key $ENCRYPTION_KEY_NAME --plaintext-file $SECRET_FILE --ciphertext-file $SECRET_FILE.encrypted
```

> ### Upload the secret

Reviewer 2: secret -> Secret

> Upload encrypted secret to GCP:

Reviewer 2: encrypted secret -> the encrypted secret

```
gsutil cp $SECRET_FILE.encrypted gs://$BUCKET_NAME/
```

> Delete the SECRET_FILE file.

Reviewer 2: Delete the file exported under the **SECRET_FILE** variable.

# Prow secrets management

## Overview

> This document explains secrets management for Prow. Some jobs require using sensitive data. You need to encrypt data using KMS and store them in GCS. The following instructions assume that you are logged-in to the Google Cloud project with administrative rights.

Reviewer 3: ```suggestion
Some jobs require using sensitive data. You need to encrypt data using KMS and store them in GCS.
This document shows the commands necessary to create a service account and store it's encrypted key in GCS bucket. The following instructions assume that you are logged-in to the Google Cloud project with administrative rights.
```

## Secrets management

->**NOTE:** Before following this guide check Prow secrets setup for the $PROJECT_NAME.

Set context to the $PROJECT_NAME executing:
```
gcloud config set project $PROJECT_NAME
```

### Create GCS bucket

The purpose of the bucket is to store encrypted credentials necessary for prow jobs like provisioning clusters/vms etc.
Run the following command:
```
gsutil mb -p $PROJECT_NAME gs://$BUCKET_NAME/
```

### Create Google Service Account

Before running following commands, export:
 - SA_NAME - Service Account name
 - SA_DISPLAY_NAME - Service Account display name
 - SECRET_FILE - the path where the private key will be written
 - ROLE - the role binded to the $SA_NAME

Create a service account:
```
gcloud iam service-accounts create $SA_NAME --display-name $SA_DISPLAY_NAME
```

Create a private key for the $SA_NAME service account:
```
gcloud iam service-accounts keys create $SECRET_FILE --iam-account=SA_NAME
```

Add policy binding for the $SA_NAME service account:
```
gcloud iam service-accounts add-iam-policy-binding $SA_NAME --member=serviceAccount:$SA_NAME@$PROJECT_NAME.iam.gserviceaccount.com --role=$ROLE
```

### Encrypt the secret

Export following:

>  - KEYRING_NAME - KMS key ring name

Reviewer 3: unnecessary
